main.rb

```diff
@@ -0,0 +1,307 @@
+
+    # my_count_block_hash
+    def h_c1(mee, counter)
+      mee.my_each { |x, y| counter += 1 if yield(x, y) }
+      counter
+    end
+
+    # my_count_parameter_array
+    def a_c2(mee, counter, par)
+      mee.my_each { |x| counter += 1 if x == par }
+      counter
+    end
+
+    # my_count_parameter_hash
+    def h_c2(mee, counter, par)
+      mee.my_each { |_x, y| counter += 1 if y == par }
+      counter
+    end
+
+    def my_count(val = nil)
+      me = self
+      c = 0
+      return me.size if val.nil? && block_given? == false
+
+      c = if me.is_a?(Array)
+            block_given? ? a_c1(me, c) { |x| yield(x) } : a_c2(me, c, val)
+          else # if me.is_a?(Hash)
+            block_given? ? h_c1(me, c) { |x, y| yield(x, y) } : h_c2(me, c, val)
+          end
+      c
+    end
+
+    def my_map_array(mee, &procs)
+      temp = []
+      i = 0
+      mee.my_each do |x|
+        temp[i] = procs.call(x)
+        temp[i] = yield(x) if block_given?
+        i += 1
+      end
+      temp
+    end
+
+    def my_map_hash(mee, &procs)
+      temp = []
+      i = 0
+      mee.my_each do |x, y|
+        temp[i] = procs.call(x, y)
+        temp[i] = yield(x, y) if block_given?
+        i += 1
+      end
+      temp
+    end
+
+    def my_map
+      me = self
+      return to_enum(:my_map) unless block_given?
+
+      if me.is_a?(Array)
+        my_map_array(me) { |x| yield(x) }
+      else # if me.is_a?(Hash)
+        my_map_hash(me) { |x, y| yield(x, y) }
+      end
+    end
+  end
+
+  # contains "my_inject" method
+  module EnumerableInside4
+    def my_inject_hash(mee)
+      temp = []
+      i = 0
+      mee.my_each do |x, y|
+        temp[i] = x
+        temp[i + 1] = y
+        i += 2
+      end
+      temp
+    end
+
+    def my_inject_array_block(mee, start = nil)
+      if start.nil?
+        start = mee.first
+        mee.my_each { |x| start = yield(start, x) if mee.first != x }
+      else
+        mee.my_each { |x| start = yield(start, x) }
+      end
+      start
+    end
+
+    def my_inject_array_proc(mee, start = nil, opr = {})
+      if start.is_a?(Symbol) # no block only second argument "operator"
+        opr = start
+        opr = opr.to_proc
+        start = mee.first
+        mee.my_each { |x| start = opr.call(start, x) if mee.first != x }
+      else
+        opr = opr.to_proc
+        mee.my_each { |x| start = opr.call(start, x) }
+      end
+      start
+    end
+
+    def my_inject(start = nil, operator = {})
+      me = self
+      # when we use hashes with inject it will convert the hash into array
+      return my_inject_hash(me) { |x, y| } unless me.is_a?(Array)
+
+      start = if block_given?
+                my_inject_array_block(me, start) { |tot, x| yield(tot, x) }
+              else
+                my_inject_array_proc(me, start, operator)
+              end
+      start
+    end
+  end
+end
+
+def multiply_els(array)
+  array.my_inject(:*)
+  # array.my_inject() {|tot, x| tot*x} # is working too
```

Comment:
* [ ] Comments: It’s better to delete commented code. If you need to review a previous version of your code, use git history ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/general/no_inline_comments.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/66243643/))
``` # array.my_inject() {|tot, x| tot*x} # is working too```